src/ejabberd_cowboy.erl

```diff
@@ -150,8 +150,9 @@ start_cowboy(Ref, Opts, Retries, SleepTime) ->
 
 -spec do_start_cowboy(atom(), listener_options()) -> {ok, pid()} | {error, any()}.
 do_start_cowboy(Ref, Opts) ->
-    #{ip_tuple := IPTuple, port := Port, handlers := Handlers,
+    #{ip_tuple := IPTuple, port := Port, handlers := Handlers0,
       transport := TransportOpts0, protocol := ProtocolOpts0} = Opts,
+    Handlers = [ Handler#{ip_tuple => IPTuple, port => Port, proto => tcp} || Handler <- Handlers0 ],
```

Comment:
Not the prettiest (listener opts mixed up with handler opts), but I guess it's the simplest option, so it's ok.